src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+        .{ "has", .KwHas },
+        .{ "where", .KwWhere },
+        .{ "implements", .KwImplements },
+        .{ "exposes", .KwExposes },
+        .{ "import", .KwImport },
+        .{ "imports", .KwImports },
+        .{ "with", .KwWith },
+        .{ "generates", .KwGenerates },
+        .{ "package", .KwPackage },
+        .{ "packages", .KwPackages },
+        .{ "requires", .KwRequires },
+        .{ "provides", .KwProvides },
+        .{ "to", .KwTo },
+        .{ "interface", .KwInterface },
+        .{ "app", .KwApp },
+        .{ "platform", .KwPlatform },
+        .{ "hosted", .KwHosted },
+    });
+
+    pub fn getKeyword(bytes: []const u8) ?T {
+        return keywords.get(bytes);
+    }
+};
+
+/// Indent holds counts for spaces and tabs.
+pub const Indent = struct {
+    num_spaces: u16,
+    num_tabs: u16,
+
+    pub fn init() Indent {
+        return Indent{ .num_spaces = 0, .num_tabs = 0 };
+    }
+
+    /// Returns:
+    /// - Some(true) if self is indented more than the other indent,
+    /// - Some(false) if less,
+    /// - null if the comparison is undefined.
+    pub fn isIndentedMoreThan(self: Indent, other: ?Indent) ?bool {
+        if (other) |o| {
+            if (self.num_spaces == o.num_spaces) {
+                return self.num_tabs > o.num_tabs;
+            } else if (self.num_tabs == o.num_tabs) {
+                return self.num_spaces > o.num_spaces;
+            } else {
+                return null;
+            }
+        } else {
+            // No minimum indent means we’re always “more indented.”
+            return true;
+        }
+    }
+};
+
+/// A simple line record: the index in the token list and its indent.
+pub const Line = struct {
+    token_start: u32,
+    indent: Indent,
+};
+
+/// The buffer that accumulates tokens.
+pub const TokenizedBuffer = struct {
+    allocator: std.mem.Allocator,
+    tokens: std.MultiArrayList(Token),
+    lines: std.ArrayList(Line),
+
+    pub fn init(allocator: std.mem.Allocator) !TokenizedBuffer {
+        return TokenizedBuffer{
+            .allocator = allocator,
+            .tokens = std.MultiArrayList(Token){},
+            .lines = std.ArrayList(Line).init(allocator),
+        };
+    }
+
+    /// Pushes a token with the given kind, token offset, and length.
+    pub fn pushToken(self: *TokenizedBuffer, kind: T, tok_offset: usize, tok_length: usize) !void {
+        try self.tokens.append(self.allocator, .{
+            .kind = kind,
+            .offset = @intCast(tok_offset),
+            .length = @intCast(tok_length),
+        });
+    }
+
+    /// Returns the offset of the token at index `idx`.
+    pub fn offset(self: *TokenizedBuffer, idx: u32) usize {
+        return @intCast(self.offsets.items[@intCast(idx)]);
+    }
+};
+
+/// A comment is represented by its begin and end offsets.
+pub const Comment = struct {
+    begin: usize,
+    end: usize,
+};
+
+/// MessageKind enumerates different diagnostic messages.
+pub const MessageKind = enum {
```

Comment:
Nit: Maybe call this ErrorKind or DiagKind or something to make it more clear that it is for handling errors.